src/megaclient.cpp

```diff
@@ -34,6 +35,80 @@ namespace mega {
 // FIXME: instead of copying nodes, move if the source is in the rubbish to reduce node creation load on the servers
 // FIXME: prevent synced folder from being moved into another synced folder
 
+
+void MegaClientAsyncQueue::push(std::function<void(SymmCipher&)> f)
+{
+    std::lock_guard<std::mutex> g(mMutex);
+    if (mAsyncThreads.empty())
+    {
+        if (f)
+        {
+            f(mZeroThreadsCipher);
+        }
+    }
+    else
+    {
+        mQueue.emplace_back(std::move(f));
+        mConditionVariable.notify_one();
+    }
+}
+
+MegaClientAsyncQueue::MegaClientAsyncQueue(Waiter& w, unsigned threadCount)
+    : mWaiter(w)
+{
+    for (int i = threadCount; i--; )
+    {
+        try
+        {
+            mAsyncThreads.emplace_back([this]()
+            {
+                asyncThreadLoop();
+            });
+        }
+        catch (std::system_error& e)
+        {
+            LOG_err << "Failed to start worker thread: " << e.what();
+            break;
+        }
+    }
+    LOG_debug << "MegaClient Worker threads running: " << mAsyncThreads.size();
+}
+
+MegaClientAsyncQueue::~MegaClientAsyncQueue()
+{
+    clearQueue();
+    push(nullptr);
+    mConditionVariable.notify_all();
+    for (auto& t : mAsyncThreads)
+    {
+        t.join();
+    }
+}
+
+void MegaClientAsyncQueue::clearQueue()
+{
+    std::lock_guard<std::mutex> g(mMutex);
+    mQueue.clear();
+}
+
+void MegaClientAsyncQueue::asyncThreadLoop()
+{
+    SymmCipher cipher;
+    std::unique_lock<std::mutex> g(mMutex);
```

Comment:
If you move the guard just before the wait then you dont need unlock/lock further down. Something like:
```cpp
for (;;)
{
    std::function<void(SymmCipher&)> f;
    {
        std::unique_lock<std::mutex> g(mMutex);
        mConditionVariable.wait(g, [this]() { return !mQueue.empty(); });
        f = std::move(mQueue.front());
        if (!f) return;
        mQueue.pop_front();
    }
    f(cipher);
    mWaiter.notify();
}
```

Reply:
I like it, nice one!  will update